Openrefine tuto addition (#5694)
galaxyproject/training-material: topics/ecology/tutorials/openrefine_gbif/tutorial.md

---
layout: tutorial_hands_on

title: Cleaning GBIF data using OpenRefine
zenodo_link: ''
questions:
- How can I use OpenRefine to clean data?
- How do I check and clean biodiversity data using OpenRefine?

objectives:

> - Use OpenRefine faceting functionnalities to apply mass editing and manage duplicates

Reviewer: ```suggestion
- Use OpenRefine faceting functionalities to apply mass editing and manage duplicates
```

> - Use OpenRefine clustering and filtering functionnalities to edit, transform data notably using regular expression

Reviewer: ```suggestion
- Use OpenRefine clustering and filtering functionalities to edit, transform data notably using regular expression
```

- Use OpenRefine to apply API services on your data
time_estimation: 0h45m
key_points:

> - OpenRefine is a powerfull tool, with many functionnalities to check, clean and enrich your data

Reviewer: ```suggestion
- OpenRefine is a powerful tool, with many functionalities to check, clean and enrich your data
```

contributions:
    authorship:
        - yvanlebras
        - sophiepamerlon
        - laurarussell
        - gbiftraining
    funding:
        - pndb

> 

Author: oh yes, THANK YOU !

---

> In this tutorial we will use OpenRefine tool to clean occurence records retrieved from GBIF.

Reviewer: ```suggestion
In this tutorial we will use OpenRefine tool to clean occurrence records retrieved from GBIF.
```

> In this tutorial we will use OpenRefine tool to clean occurence records retrieved from GBIF.

Reviewer: you sometimes write OpenRefine, and sometimes Openrefine (without capitalization of the R), best to pick one and be consistent

> This tutorial is based on the GBIF {% cite GBIFtraining2021 %} "Biodiversity data mobilization course" accessible on [GBIF website](https://docs.gbif.org/course-data-mobilization/en/) on Use case 1 ["Herbarium Specimens"](https://docs.gbif.org/course-data-mobilization/en/exercise-3a-c.html) using file from Exercice 3C, at the bottom of the page, reachable through this direct [URL](https://docs.gbif.org/course-data-mobilization/exercise-data/UC1-3c-open-refine.csv).

Reviewer: ```suggestion
This tutorial is based on the GBIF {% cite GBIFtraining2021 %} "Biodiversity data mobilization course" accessible on [GBIF website](https://docs.gbif.org/course-data-mobilization/en/) on Use case 1 ["Herbarium Specimens"](https://docs.gbif.org/course-data-mobilization/en/exercise-3a-c.html) using file from Exercise 3C, at the bottom of the page, reachable through this direct [URL](https://docs.gbif.org/course-data-mobilization/exercise-data/UC1-3c-open-refine.csv).
```

Author: THNAK YOU eagle eye !

> <agenda-title></agenda-title>
>
> In this tutorial, we will cover:
>

> > 1. Import data to OpenRefine
> > 2. Faceting
> > 3. Filtering
> > 4. Clustering
> > 5. API use

Reviewer: The agenda will be autogenerated based on your sections, please don't manually edit. If you want the agenda to look a certain way, may your sections reflect this structure (e.g. all 1st and 2nd level, so `#` and `##` will show up in the agenda)

```suggestion

```

> {:toc}
>
{: .agenda}

# Retrieve data from GBIF "Biodiversity data mobilization course"

## Get data

> <hands-on-title> Data import and project creation </hands-on-title>
>
> 1. Create a new history for this tutorial and give it a name (example: “Openrefine tutorial”) for you to find it again later if needed.
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
>    {% snippet faqs/galaxy/histories_rename.md %}
>
> 2. Import data file from GBIF {% cite GBIFtraining2021 %} "Biodiversity data mobilization course" accessible on [GBIF website](https://docs.gbif.org/course-data-mobilization/en/) on Use case 1 ["Herbarium Specimens"](https://docs.gbif.org/course-data-mobilization/en/exercise-3a-c.html) using file from Exercice 3C, at the bottom of the page
>
>    "Herbarium Specimens" data file as a csv file named "UC1-3c-open-refine.csv"
>    ```

> >    https://docs.gbif.org/course-data-mobilization/exercise-data/UC1-3c-open-refine.csv

Author: Hmm, like for a previous tutorial, I really prefer to not duplicate data and here, the "paternity" of the dataset is not easy to display if I upload it myself on a data warehouse.... I prefer to let this link for now if possible, so it is clear that the dataset is accessible through GBIF training website... PArallely, I can try to have direct discussion with GBIF colleague to know if they can push it on Zenodo.

> >    https://docs.gbif.org/course-data-mobilization/exercise-data/UC1-3c-open-refine.csv

Reviewer: It would be great to at least add it to Zenodo as a backup. The instructions here can still point to gbif website if that is a useful resource for learners to know about, but by adding to Zenodo we get 1) a backup in case this URL ever changes, and 2) our automations to add the file to a galaxy data library will work.

So just adding it to Zenodo and putting the zenodo link in metadata would be enoug, this box can stay unchanged.

> >    https://docs.gbif.org/course-data-mobilization/exercise-data/UC1-3c-open-refine.csv

Author: ok, I just sent a message to GBIF colleagues to see what they think is doable.

>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
>
> 3. {% tool [Regex Find And Replace](toolshed.g2.bx.psu.edu/repos/galaxyp/regex_find_replace/regex1/1.0.3) %} with the following parameters:
>    - {% icon param-file %} *"Select lines from"*: `UC1-3c-open-refine.csv`
>    - In *"Check"*:
>        - {% icon param-repeat %} *"Insert Check"*
>            - *"Find Regex"*: `;`
>            - *"Replacement"*: `\t`
> 4. Change the datatype from "txt" to "tabular"
> 
>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="tabular" %}
> 
> 6. Check your output. All columns must be separated by a tabulation now. See example below.
>

> > ![Output of Regex Find And Replace example](../../images/openrefine/1_openrefine.png "Output of Regex Find And Replace example")

Author: THANK YOU Saskia !

> 
>
> 7. Rename the datasets if needed, for example with "UC1-3c-open-refine-tabular" to keep orginal name and specify the tabular format.
>
>    {% snippet faqs/galaxy/datasets_rename.md %}
>
> 
>
{: .hands_on}

## Deploy an Openrefine instance and push your data in

> <hands-on-title> Deploy an Openrefine instance and import data from your Galaxy history </hands-on-title>
>

> > 1. Deploy an OpenRefine instance clicking Execute on the {% tool [OpenRefine tool](interactive_tool_openrefine) %} specifying "UC1-3c-open-refine-tabular" as input dataset

Reviewer: Please use the GTN formatting for hands-on instructions:

```suggestion
> 1. {% tool [OpenRefine tool](interactive_tool_openrefine) %} with the following parameters:
>    - *"Input file in tabular format"*: `UC1-3c-open-refine-tabular`
```

Author: THANK YOU !

> > 2. You can then go n the OpenRefine instance, and go to "Open Project" space, then select the project named "Galaxy file"

Reviewer: This is one sentence with 3 steps, for clarity I would suggest breaking these up in bullet points, with clear concise actions for the user to take 

```suggestion
> 2. Open the OpenRefine instance
     - go to "Open Project" space
     - select the project named "Galaxy file"
```

Author: THANK YOU !!

> 3. You will now have your project with first 10 lines of your file displayed
> ![Openrefine project start](../../images/openrefine/openrefine_start.PNG)
>
{: .hands_on}

## Faceting

Faceting is a feature that will allow us to get a big picture overview of the data, and to filter down to just the subset of rows
that we want to change or view in bulk. It facilitates the use and analysis of data and can be done with cells containing any
kind of text, numbers and dates.

> <hands-on-title> Hands-on: Faceting and mass editing </hands-on-title>
>
> 1. Go to column "kingdom", and then click on the column menu and follow the route to `Text facet`.
>    - On the left a window with the name of the column will appear, that is the facet.
> ![Openrefine facet kingdom](../../images/openrefine/openrefine_facet_kingdom.PNG)
> 2. Click on count to sort by count, then click on name to sort alphabetically
> 3. Fix the spelling mistakes (Plante -> Plantae). Place the cursor over the text in the window and click on `edit`, then fix the error in the text box, and to save click on apply.
>    - All the values will be fixed automatically.
>
>

> >    > <comment-title></comment-title>

Reviewer: empty comment box?

>    - search for "spp".
>       - Go to "Full name" column menu, then click `Edit cells > Transform`
>       - In the text box paste the formula `value.replace(" spp.", "")` and click `ok`
>
{: .hands_on}

> <question-title></question-title>
>
> 1. Do you think you can use also `Edit cells > Transform` formula to apply the changes you made manually for the 2 first terms searched ("sp1" and "SP2")?
> 2. Why you can easily edit directly these 2 first entries and not the third one?
>
> > <solution-title></solution-title>
> >
> > 1. Yes, using respectively `value.replace(" sp1", "")` and `value.replace(" SP2", "")` formulas
> > 2. Because the 2 first searchs only concern one entry for each when the third one return several.
> >
> {: .solution}
>
{: .question}

> <hands-on-title> {% icon hands_on %} Hands-on: Advance filter 1. </hands-on-title>
>
> 1. Go to column "genus" and perform a `Text filter`.
> 2. Check `regular expression` and `case sensitive` , then paste the expression "^[a-z]"
>    - This regular expression filters the strings in which the first letter is lowercase.

> > 3. Perform a correction since the genus should be capitalized.

Reviewer: you could add our Regular Expression FAQ in this box if you like? https://training.galaxyproject.org/training-material/faqs/galaxy/analysis_regular_expressions.html

>
>
>    > <comment-title></comment-title>
>    >

> >    > If you want to know more about regular expressions, you can consult https://openrefine.org/docs/manual/expressions#regular-expressions and specifically for GREL functions: https://openrefine.org/docs/manual/grelfunctions

Reviewer: (These  URLs weren't being rendered as clickable)

```suggestion
>    > If you want to know more about regular expressions, you can consult the [Openrefine regular expressions documentation](https://openrefine.org/docs/manual/expressions#regular-expressions) and specifically the [documentation for GREL function](https://openrefine.org/docs/manual/grelfunctions)
```

>    - The facet will show 4 duplicates
> 2. Click on true, and you'll see the values in the main window
> 3. After a check with the specimens labels, fix the values clicking edit directly on the cell with the correct catalogue numbers
>    - UWP:122470 Vargas P
>    - UWP:122471 Vargas I
>    - UWP:157351 Betancur H
>    - UWP:157339 Betancur J
>
{: .hands_on}

## Filtering 

> <hands-on-title> {% icon hands_on %} Hands-on: Basic filter. </hands-on-title>
>
> 1. Go again to "Full name" column menu and perform a `Text facet` to visualize the values, then go again to the column menu and click on `Text filter`, perform the following filters and fix them as described below:
>    - search for "sp1" entries -> Then remove it and obtain "Cyperus", clicking on `Edit` directly in the cell
>    - search for "SP2" entries, check `case sensitive` -> To remove it and obtain "Cyperus", you can `Edit` directly in the cell
>    - search for "spp".
>    > 1. Go to "Full name" column menu, then click `Edit cells > Transform`
>    > 2. In the text box paste the formula `value.replace(" spp.", "")` and click `ok`
>
{: .hands_on}

> > ### {% icon question %} Questions

Author: THANK YOU, that's what I was starting to understand ;) I am looking at it!

---
layout: tutorial_hands_on

title: Cleaning GBIF data using OpenRefine
zenodo_link: ''
questions:
- How can I use OpenRefine to clean data?
- How do I check and clean biodiversity data using OpenRefine?

objectives:
- Use OpenRefine faceting functionnalities to apply mass editing and manage duplicates
- Use OpenRefine clustering and filtering functionnalities to edit, transform data notably using regular expression
- Use OpenRefine to apply API services on your data
time_estimation: 0h45m
key_points:
- OpenRefine is a powerfull tool, with many functionnalities to check, clean and enrich your data

> contributors:

Reviewer: @yvanlebras  this should fix the linting error

```suggestion
contributions:
```

Author: Oh! THANK YOU so much Saskia!
